test/functional/mempool_rebroadcast.py

```diff
@@ -0,0 +1,289 @@
+        fee_rate = mempool_entry['fee']*1000/mempool_entry['vsize']
+        assert_approx(float(fee_rate), float(desired_fee_rate))
+
+        return tx_hsh
+
+    def test_simple_rebroadcast(self):
+        self.log.info("Test simplest rebroadcast case")
+
+        node0 = self.nodes[0]
+        node1 = self.nodes[1]
+
+        # generate mempool transactions that both nodes know about
+        for _ in range(3):
+            node0.sendtoaddress(node1.getnewaddress(), 4)
+
+        self.sync_all()
+
+        # generate mempool transactions that only node0 knows about
+        disconnect_nodes(node0, 1)
+
+        for _ in range(3):
+            node0.sendtoaddress(node1.getnewaddress(), 5)
+
+        # check that mempools are different
+        assert_equal(len(node0.getrawmempool()), 6)
+        assert_equal(len(node1.getrawmempool()), 3)
+
+        # reconnect the nodes
+        connect_nodes(node0, 1)
+
+        # rebroadcast will only occur if there has been a block since the
+        # last run of CacheMinRebroadcastFee. when we connect a new peer, rebroadcast
+        # will be skipped on the first run, but caching will trigger.
+        # have node1 generate so there are still mempool txns that need to be synched.
+        node1.generate(1)
+
+        assert_equal(len(node1.getrawmempool()), 0)
+        wait_until(lambda: len(node0.getrawmempool()) == 3)
+
+        # bump time to hit rebroadcast interval
+        mocktime = int(time.time()) + 300 * 60
+        node0.setmocktime(mocktime)
+        node1.setmocktime(mocktime)
+
+        # check that node1 got txns bc rebroadcasting
+        wait_until(lambda: len(node1.getrawmempool()) == 3, timeout=30)
+
+        global global_mocktime
+        global_mocktime = mocktime
+
+    def test_recency_filter(self):
+        self.log.info("Test recent txns don't get rebroadcast")
+
+        node = self.nodes[0]
+        node1 = self.nodes[1]
+
+        global global_mocktime
+        mocktime = global_mocktime
+
+        node.setmocktime(global_mocktime)
+        node1.setmocktime(global_mocktime)
+
+        # mine blocks to clear out the mempool
+        node.generate(4)
+        assert_equal(len(node.getrawmempool()), 0)
+
+        # add p2p connection
+        conn = node.add_p2p_connection(P2PTxInvStore())
+
+        # create old txn
+        node.sendtoaddress(node.getnewaddress(), 2)
+        assert_equal(len(node.getrawmempool()), 1)
+        wait_until(lambda: conn.get_invs(), timeout=30)
+
+        # bump mocktime to ensure the txn is old
+        mocktime += 31 * 60 # seconds
+        node.setmocktime(mocktime)
+
+        delta_time = 28 * 60 # seconds
+        while True:
+            # create a recent transaction
+            new_tx = node1.sendtoaddress(node1.getnewaddress(), 2)
+            new_tx_id = int(new_tx, 16)
+
+            # ensure node0 has the transaction
+            wait_until(lambda: new_tx in node.getrawmempool())
```

Comment:
This assertion timed out for me ("not true after 60 seconds") on the first run of 375 secs. Passed on a second run, then timed out in the same place again in the third run.  Passed a few more runs, then failed again.

Reply:
Both at 42c460a24f6d9d9c2c65c30287965acbce096ecf and rebased to current master.

```
~/projects/bitcoin/bitcoin (pr/16698) $ test/functional/mempool_rebroadcast.py 
2019-11-27T17:51:11.955000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_g9hdr2e_
2019-11-27T17:51:13.096000Z TestFramework (INFO): Test simplest rebroadcast case
2019-11-27T17:51:14.928000Z TestFramework (INFO): Test recent txns don't get rebroadcast
2019-11-27T17:56:31.453000Z TestFramework.utils (ERROR): wait_until() failed. Predicate: ''''
            wait_until(lambda: new_tx in node.getrawmempool())
'''
2019-11-27T17:56:31.467000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 111, in main
    self.run_test()
  File "test/functional/mempool_rebroadcast.py", line 44, in run_test
    self.test_recency_filter()
  File "test/functional/mempool_rebroadcast.py", line 171, in test_recency_filter
    wait_until(lambda: new_tx in node.getrawmempool())
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/util.py", line 229, in wait_until
    raise AssertionError("Predicate {} not true after {} seconds".format(predicate_source, timeout))
AssertionError: Predicate ''''
            wait_until(lambda: new_tx in node.getrawmempool())
''' not true after 60 seconds
2019-11-27T17:56:31.519000Z TestFramework (INFO): Stopping nodes
```

Reply:
This on my first run as well, but then passed six in a row so didn't report.

Reply:
I'm unable to reproduce this locally. Asked @jonatack for logs & looks like the failure occurs after going through the while loop >150 times, which means the `if new_conn.get_invs()` conditional is never hitting. Theoretically it would be possible for the condition to not hit, but that should be an _extreme_ edge case, vs the frequency of failures being seen. I'll try to run this test in different environments to see if I can reproduce.

Reply:
FWIW this was on a recent Debian and building with `--enable-debug --enable-bench`. Could be good to have other data points in the form of people running this test. I hope to dig into these tests soon.

Reply:
I was able to reproduce this failure on an ubuntu machine. latest push fixed it for me & jonatack. going to resolve.

Reply:
> I was able to reproduce this failure on an ubuntu machine. latest push fixed it for me & jonatack. going to resolve.

SGTM